### PR TITLE
[SPARK-8398][CORE] Hadoop input/output format advanced control

### PR DESCRIPTION
**Pull request by koertkuipers**
## What changes were proposed in this pull request?

Consistently expose Configuration/JobConf for all methods that use Hadoop input/output formats, which facilitates re-use and discourages many additional parameters (that end up changing the Configuration/JobConf internally).
## How was this patch tested?

New tests in SparkContextSuite that check that the resulting HadoopRDD/NewHadoopRDD indeed has the settings passed in using the Configuration/JobConf parameter.

### CODE REVIEW

**Comment by squito** [comment]
Jenkins, this is OK to test

**Comment by squito** [comment]
I think we'll also want to add these to `JavaSparkContext` and the python api (though I am not entirely certain how to do that off the top of my head.)
It would also be nice to have some really simple regression tests that those confs actually get used.

**Comment by koertkuipers** [comment]
ok i will look into JavaSparkContext and a few simple regression tests.
will probably need some help with python.

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ import scala.reflect.{classTag, ClassTag}
 import com.clearspring.analytics.stream.cardinality.HyperLogLogPlus
 import org.apache.hadoop.io.{BytesWritable, NullWritable, Text}
 import org.apache.hadoop.io.compress.CompressionCodec
-import org.apache.hadoop.mapred.TextOutputFormat
+import org.apache.hadoop.mapred.{ JobConf, TextOutputFormat }
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: remove space around the {}

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 2]
```diff
@@ -1397,7 +1397,7 @@ abstract class RDD[T: ClassTag](
       }
     }
     RDD.rddToPairRDDFunctions(r)(nullWritableClassTag, textClassTag, null)
-      .saveAsHadoopFile[TextOutputFormat[NullWritable, Text]](path)
+      .saveAsHadoopFile(path, classOf[NullWritable], classOf[Text], classOf[TextOutputFormat[NullWritable, Text]], conf)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: this line is almost certainly too long. Can you break it at `(`?

File: core/src/main/scala/org/apache/spark/rdd/RDD.scala [hunk 3]
```diff
@@ -1421,10 +1421,10 @@ abstract class RDD[T: ClassTag](
   /**
    * Save this RDD as a SequenceFile of serialized objects.
    */
-  def saveAsObjectFile(path: String): Unit = withScope {
+  def saveAsObjectFile(path: String, conf: JobConf = new JobConf(sc.hadoopConfiguration)): Unit = withScope {
     this.mapPartitions(iter => iter.grouped(10).map(_.toArray))
       .map(x => (NullWritable.get(), new BytesWritable(Utils.serialize(x))))
-      .saveAsSequenceFile(path)
+      .saveAsSequenceFile(path, None, conf)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rdd/RDD.scala]: nit: I would do `saveAsSequenceFile(path, conf = conf)` instead

**Comment by andrewor14** [comment]
add to whitelist

**Comment by andrewor14** [comment]
The changes here look fine. @JoshRosen do we have to worry about breaking binary compatibility in some ways here? Even though we provide a default value to the last parameter we're technically adding a new parameters to a several public APIs here.

**Comment by JoshRosen** [comment]
@andrewor14 Adding a new parameter with a default value will break binary compatibility from a Java point-of-view, as far as I know.

**Comment by JoshRosen** [comment]
MiMa should tel us, though.

**Comment by koertkuipers** [comment]
i see MiMa failed. what binary compatibility promise does spark make? all minor versions are binary compatible?

**Comment by andrewor14** [comment]
retest this please. MiMa tests have been a little flaky recently.

**Comment by koertkuipers** [comment]
i see:
[info] spark-core: found 2 potential binary incompatibilities (filtered 488)
[error]  * method saveAsTextFile(java.lang.String,org.apache.hadoop.mapred.JobConf)Unit in trait org.apache.spark.api.java.JavaRDDLike does not have a correspondent in old version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.api.java.JavaRDDLike.saveAsTextFile")
[error]  * method saveAsObjectFile(java.lang.String,org.apache.hadoop.mapred.JobConf)Unit in trait org.apache.spark.api.java.JavaRDDLike does not have a correspondent in old version
[error]    filter with: ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.api.java.JavaRDDLike.saveAsObjectFile")

the issue seems to be that i am introducing new methods. is that about forward compatibility? not sure what to do here.

**Comment by andrewor14** [comment]
Usually you can just copy and paste whatever the MiMa checker tells you to add to `MimaExcludes.scala`. For instance, add `ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.api.java.JavaRDDLike.saveAsTextFile")` to that file.

**Comment by koertkuipers** [comment]
i will fix resolve conflicts when someone says this is good to go. otherwise i keep merging from master every few days.

**Comment by koertkuipers** [comment]
@andrewor14 @JoshRosen anything i need to do, besides fixing trivial conflicts?

**Comment by holdenk** [comment]
Would we want to maybe consider this for Spark 2.0? It seems like if were maybe going to be adding new default params to functions this might be the time to do it (of course only if people have the bandwidth to update & also review)?

It also seems like some unrelated R changes might have accidentally gotten mixed in during one of the merges that should be reverted if we want to move forward with this.

**Comment by koertkuipers** [comment]
i am happy to update this, if there is any interest. or otherwise i will
close.

**Comment by ScrapCodes** [comment]
IMO, this is useful in one way that hadoop configuration need not be a global state. We can have a default set of configuration that we use everywhere as a default. And then in every hadoop related method a user has an alternative to override the default. 

Binary compatibility will definitely be broken, but source compatibility might not be affected i.e. one might need to recompile the project with newer spark version. As it is asked already, it should be okay for 2.0 ?

@andrewor14 ping !

**Comment by koertkuipers** [comment]
Jenkins, retest this please.

**Comment by koertkuipers** [comment]
ok i updated this for spark 2. the unit test failures seem unrelated

**Comment by ScrapCodes** [comment]
Jenkins, retest this please.

**Comment by holdenk** [comment]
@koertkuipers now days we try and provide a description for our pull request (sometimes it can be copied from the JIRA) for the eventual commit message - it might be good to add that?
Also CC @JoshRosen & @squito to maybe take a look at the updated PR :)

**Comment by koertkuipers** [comment]
@holdenk ok i tried to make it look all pretty

**Comment by rxin** [comment]
I'm going to close this for now.

**Pull request closed without merging**